packages/core-data/src/entities.ts

```diff
@@ -12,157 +12,296 @@ import { __ } from '@wordpress/i18n';
 		kind: 'root',
 		baseURL: '/wp/v2/comments',
 		baseURLParams: { context: 'edit' },
 		plural: 'comments',
 		label: __( 'Comment' ),
-	},
-	{
+	} as const ),
+	defineEntity( {
 		name: 'menu',
 		kind: 'root',
 		baseURL: '/wp/v2/menus',
 		baseURLParams: { context: 'edit' },
 		plural: 'menus',
 		label: __( 'Menu' ),
-	},
-	{
+	} as const ),
+	defineEntity( {
 		name: 'menuItem',
 		kind: 'root',
 		baseURL: '/wp/v2/menu-items',
 		baseURLParams: { context: 'edit' },
 		plural: 'menuItems',
 		label: __( 'Menu Item' ),
 		rawAttributes: [ 'title', 'content' ],
-	},
-	{
+	} as const ),
+	defineEntity( {
 		name: 'menuLocation',
 		kind: 'root',
 		baseURL: '/wp/v2/menu-locations',
 		baseURLParams: { context: 'edit' },
 		plural: 'menuLocations',
 		label: __( 'Menu Location' ),
 		key: 'name',
-	},
-	{
+	} as const ),
+	defineEntity( {
 		name: 'navigationArea',
 		kind: 'root',
 		baseURL: '/wp/v2/block-navigation-areas',
 		baseURLParams: { context: 'edit' },
 		plural: 'navigationAreas',
 		label: __( 'Navigation Area' ),
 		key: 'name',
-		getTitle: ( record ) => record?.description,
-	},
-	{
+		getTitle: ( record: EntityTypes.NavigationArea< 'edit' > | null ) =>
+			record?.description,
+	} as const ),
+	defineEntity( {
 		label: __( 'Global Styles' ),
 		name: 'globalStyles',
 		kind: 'root',
 		baseURL: '/wp/v2/global-styles',
 		baseURLParams: { context: 'edit' },
 		plural: 'globalStylesVariations', // should be different than name
 		getTitle: ( record ) => record?.title?.rendered || record?.title,
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Themes' ),
 		name: 'theme',
 		kind: 'root',
 		baseURL: '/wp/v2/themes',
 		baseURLParams: { context: 'edit' },
 		key: 'stylesheet',
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Plugins' ),
 		name: 'plugin',
 		kind: 'root',
 		baseURL: '/wp/v2/plugins',
 		baseURLParams: { context: 'edit' },
 		key: 'plugin',
-	},
-];
+	} as const ),
+] as const;
+
+type Element = typeof defaultEntities[ number ];
+
+type DefaultEntityContext<
+	K extends string,
+	N extends string,
+	E = Extract< Element, { kind: K; name: N } >
+> = 'baseURLParams' extends keyof E
+	? 'context' extends keyof E[ 'baseURLParams' ]
+		? E[ 'baseURLParams' ][ 'context' ]
+		: 'view'
+	: 'view';
+
+type EntityKeyName<
+	K extends string,
+	N extends string,
+	E = Extract< Element, { kind: K; name: N } >
+> = 'key' extends keyof E ? E[ 'key' ] : 'id';
+
+type DeclaredEntity< K extends string, N extends string, RecordType > = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: EntityKeyName< K, N > extends keyof RecordType
+		? RecordType[ EntityKeyName< K, N > ]
+		: never;
+	defaultContext: DefaultEntityContext< K, N >;
+};
+type APIEntity<
+	K extends string,
+	N extends string,
+	RecordType,
+	KeyName = 'id',
+	C = 'edit'
+> = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: KeyName extends keyof RecordType ? RecordType[ KeyName ] : never;
+	defaultContext: C;
+};
```

Comment:
what's the difference between a declared entity and an API entity and when and why should I choose one over the other?

Reply:
Some configuration data is hardcoded in `entities.js`: https://github.com/WordPress/gutenberg/blob/a8ea1078c42a3312ee5402c2316086b50297a671/packages/core-data/src/entities.js#L21-L30

Other is loaded by requesting the API:
https://github.com/WordPress/gutenberg/blob/a8ea1078c42a3312ee5402c2316086b50297a671/packages/core-data/src/entities.js#L207

With that in mind:
 
* DeclaredEntity is a name I came up with for the ones with a hardcoded configuration that we can directly reuse.
* APIEntity is a name for the entities with no TS-reusable configuration where we must explicitly specify the `defaultContext` and `key`.

Reply:
is the distinction important? Do we need it at all?

Reply:
I elaborated on that in https://github.com/WordPress/gutenberg/pull/39025#issuecomment-1054327402